src/js/pouchdb-common.js

```diff
@@ -0,0 +1,145 @@
+/*
+
+    Provide a component that wraps an individual PouchDB instance.
+
+    https://github.com/GPII/gpii-pouchdb/blob/master/docs/pouchdb.md
+
+ */
+/* eslint-env node */
+"use strict";
+var fluid = fluid || require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+
+var PouchDB = PouchDB || require("pouchdb");
+
+fluid.registerNamespace("gpii.pouch");
+
+gpii.pouch.init = function (that) {
+    that.pouchDb = new PouchDB(that.options.dbOptions);
+    // fluid.log("Pouch instance `" + that.options.dbOptions.name + "` (" + that.id + ") initialized...");
+};
+
+/**
+ *
+ * Call an underlying PouchDb function and fire the named function on completion.
+ *
+ * @param that - The component itself.
+ * @param fnName {String} - The name of the PouchDB function to call.
+ * @param fnArgs {Array} - The arguments (minus the final callback) to pass to the function.
+ * @param eventName {String} - The event to fire on completion.
+ *
+ */
+gpii.pouch.callPouchFunction = function (that, fnName, fnArgs, eventName) {
+    var promise = fluid.promise();
+
+    var wrappedCallback = function (err, results) {
+        if (err) {
+            promise.reject(err);
+            that.events.onError.fire(err);
+        }
+        else {
+            promise.resolve(results);
+            that.events[eventName].fire(results);
+        }
+    };
+
+    // Add the event firer callback to the list of arguments as the final argument.  If there are no args, we create
+    // an empty array first.
+    var fullArgs = fluid.makeArray(fnArgs).concat(wrappedCallback);
+    that.pouchDb[fnName].apply(that.pouchDb, fullArgs);
+
+    return promise;
+};
+
+fluid.defaults("gpii.pouch", {
+    gradeNames: ["fluid.component"],
+    dbOptions: {
+        skip_setup: true
+    },
+    events: {
+        onAllDocsComplete: null,
```

Comment:
I'd written a response that read like this: 

In fact this only gives an impression of safety. Replacing a promise-based API one for one with an event-based one creates an unavoidable risk of "cross-contamination" between requests. Having more event types reduces the contamination but does not eliminate it. So we may as well be honest that these events are only intended for the lightweight use of the IoC testing framework. I've written up https://issues.fluidproject.org/browse/FLUID-5949 to describe these kinds of risks further, and propose an improvement in the testing framework that will remove the encouragement to create these kinds of event-based wrappers.

I thought further, and realised that I don't fully believe in this advice. There *do* seem to be some legitimate use cases for allowing the user to intercept these kinds of events by type, for example, to interpose special behaviour or processing for example, on "onCompactComplete" etc.

However, you will see my point about the risks of contamination - there's no way that the consumer of the onGetComplete event can know *which* get is complete. There is nothing which associates an particular invocation of {pouch}.get with a particular firing of {pouch}.events.onGetComplete - in test sequences we tend to fudge by assuming that the *next* invocation of onGetComplete has to be the matching one, but it's clear that in real use this is faulty.

So what I propose instead is this - that you eliminate all of these invokers and all of these events, and move instead to the "transforming promise sequence" pattern that we see in DataSource. "get" is now signalled by that.get.fire, and the "implementation" of the invoker just becomes another listener to this event. It's then possible i) for users to interpose both before and after the actual dispatch to Pouch, and ii) the ground is opened for a future generic "request id" system that dumps an extra value into the "options" structure of fluid.promise.sequence iii) we no longer have the annoying pleonasm of a whole raft of invokers and events with corresponding but not completely matching names

In practice we plan to abolish invokers completely, and we may as well start experimenting with this technique in earnest and develop infrastructure to support it. I will also try to take care of https://issues.fluidproject.org/browse/FLUID-5949 as soon as possible